scripts/ansible/runner.py

```diff
@@ -0,0 +1,257 @@
+            ansible_hosts.append(play["hosts"])
+
+    if not ansible_hosts:
+        logger.error("No playbook hosts found")
+        sys.exit(1)
+
+    logger.info(f"Playbook Hosts Found: {ansible_hosts}")
+
+    return ansible_hosts
+
+
+def print_images(images: List[Image]):
+    print("\nAvailable Images:")
+    for image in images:
+        print(image)
+
+
+vms: List[VM] = []
+
+
+def cleanup(*args, **kwargs):
+    if not vms:
+        return
+
+    logger.debug(f"Running cleanup")
+    while vms:
+        vm = vms.pop()
+        logger.debug(f"Destroying {vm}")
+        vm.destroy()
+
+
+def modified(host: Host, image: Image) -> bool:
+    """
+    Return True if the playbook is modified. Otherwise, return False.
+
+    A playbook is considered modified if its modified timestamp is newer than
+    the modified timestamp of the image at the export path.
+
+    This avoids updating XVAs that are already up-to-date.
+    """
+    playbook_modified = int(os.stat(args.playbook).st_mtime)
+
+    if not host.file_exists(image.export_path):
+        return True
+
+    image_modified = int(host.ssh(["stat", "--format", "%Y", image.export_path]))
+    if image_modified > playbook_modified:
+        logger.info((
+            f"Skipping {image.image_name}, the file found at "
+            f"{image.export_path} is newer than the playbook"
+        ))
+        return False
+
+    return True
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Run a playbook for updating test VMs")
+    parser.add_argument(
+        "--http",
+        type=str,
+        required=False,
+        default=getattr(data, "DEF_VM_URL", "http://pxe/images/"),
+        help=(
+            "The HTTP/PXE server containing the test VMs. Defaults to DEF_VM_URL "
+            "in data.py, if not found then defaults to http://pxe/images/"
+        ),
+    )
+    parser.add_argument(
+        "--forced",
+        action="store_true",
+        help="Force update all VMs. Don't skip any due to modified timestamps."
+    )
+    parser.add_argument(
+        "--loglevel",
+        choices=["DEBUG", "ERROR", "WARNING", "INFO"],
+        default="INFO",
+        help="Force update all VMs. Don't skip any due to modified timestamps."
+    )
+    parser.add_argument(
+        "--print-images",
+        action="store_true",
+        required=False,
+        help="Show the available images on the PXE server and then exit.",
+    )
+
+    # Args host/playbook are not required if the user only wants to print the images
+    if "--print-images" not in sys.argv:
+        parser.add_argument(
+            "--host",
+            "-x",
+            type=str,
+            help="The XCP-ng hostname or IP address. Defaults to host found in data.py",
+        )
+        parser.add_argument(
+            "--export-directory",
+            "-e",
+            type=str,
+            required=False,
+            default=os.path.join("/", "root", "ansible-updates/"),
```

Comment:
This might be feasible for XCP-ng at large, but would maybe be best to have this resolve to the user's home.